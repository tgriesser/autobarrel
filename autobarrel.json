{
  "paths": ["src/**"],
  "exclude": ["src/cli.ts"]
}
