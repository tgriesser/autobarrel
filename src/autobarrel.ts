import chokidar from "chokidar"
import * as fs from "fs"
import glob from "glob"
import * as path from "path"
import util from "util"

const statAsync = util.promisify(fs.stat)
const writeFileAsync = util.promisify(fs.writeFile)
const readFileAsync = util.promisify(fs.readFile)
const globAsync = util.promisify(glob)

export type AutoBarrelConfig = {
  /**
   * Absolute path to autobarrel.json
   */
  path: string
}

export type AutoBarrelData = {
  /**
   * Absolute glob patterns we want to process
   */
  paths: string[]
  /**
   * Absolute paths we want to ignore entirely
   */
  ignore?: string[]
  /**
   * Absolute paths of files / directories we want to avoid re-exporting
   */
  exclude?: string[]
}

export async function autobarrel(configData: AutoBarrelData) {
  const promises = []
  const globbedFiles = new Set<string>()
  const excludeFiles = new Set<string>()

  promises.push(
    ...(configData.exclude?.map(async (p) => {
      const exclude = await globAsync(p, {
        ignore: configData.ignore || [],
      })
      exclude.map((e) => {
        if (path.extname(e) === "") {
          excludeFiles.add(e + "/index.ts")
        } else {
          excludeFiles.add(e)
        }
      })
    }) ?? [])
  )

  promises.push(
    ...configData.paths.map(async (p) => {
      const added = await globAsync(p, {
        ignore: configData.ignore || [],
      })
      added.map((a) => globbedFiles.add(a))
    })
  )

  await Promise.all(promises)

  const toBarrel = Array.from(globbedFiles)
  const toBarrelMap: Record<string, Set<string>> = {}
  const dirCandidates: string[] = []
  const tsFiles = new Set<string>()
  const directories = new Set<string>()

  toBarrel.forEach((p) => {
    if (p.match(/\.tsx?$/) && !p.endsWith(".d.ts")) {
      tsFiles.add(p)
    } else if (!path.extname(p)) {
      dirCandidates.push(p)
    }
  })

  await Promise.all(
    dirCandidates.map(async (dir) => {
      const stat = await statAsync(dir)
      if (stat.isDirectory()) {
        directories.add(dir)
        const barrelFile = path.join(dir, "index.ts")
        tsFiles.add(barrelFile)
        if (excludeFiles.has(dir)) {
          excludeFiles.add(barrelFile)
        }
      }
    })
  )

  Array.from(directories).forEach((dir) => {
    toBarrelMap[dir] = new Set()
  })

  Array.from(tsFiles).forEach((file) => {
    const dirName = path.dirname(file)
    if (
      toBarrelMap[dirName] &&
      (!excludeFiles.has(file) || path.basename(file) === "index.ts")
    ) {
      toBarrelMap[dirName].add(file)
    }
  })

  Object.keys(toBarrelMap).forEach((k) => {
    const dirName = path.dirname(k)
    if (
      toBarrelMap[dirName] &&
      toBarrelMap[k].has(path.join(k, "index.ts")) &&
      !excludeFiles.has(path.join(k, "index.ts"))
    ) {
      toBarrelMap[dirName].add(k)
    }
  })

  Object.entries(toBarrelMap).forEach(([key, val]) => {
    maybeDestroy(key, val)
  })

  function maybeDestroy(key: string, val: Set<string>) {
    if (toBarrelMap[key]) {
      if (val.size === 1) {
        delete toBarrelMap[key]
        const p = path.dirname(key)
        if (toBarrelMap[p]) {
          toBarrelMap[p].delete(key)
          maybeDestroy(p, toBarrelMap[p])
        }
      }
    }
  }

  const pathsWritten = await Promise.all(
    Object.entries(toBarrelMap).map(async ([key, val]) => {
      const lines = ["// created by autobarrel, do not modify directly\n"]
      const toAdd = Array.from(val).sort()

      toAdd
        .map((p) => path.relative(key, p))
        .map((p) => (path.extname(p) ? p : `${p}/`))
        .forEach((p) => {
          if (path.basename(p) !== "index.ts") {
            lines.push(`export * from './${p.replace(/\.tsx?$/, "")}'`)
          }
        })
      const writePath = path.join(key, "index.ts")
      await writeFileAsync(writePath, lines.join("\n"))
      return writePath
    })
  )

  return pathsWritten.sort()
}

export type AutoBarrelWatchData = AutoBarrelData & {
  /**
   * Whether the watch is running in verbose mode
   */
  verbose?: boolean
}

export async function autobarrelWatch(configData: AutoBarrelWatchData) {
  let running = false
  await autobarrel(configData)
  const watcher = chokidar.watch(configData.paths, {
    ignoreInitial: true,
    ignored: configData.ignore ?? [],
  })
  async function maybeAutobarrel(p: string) {
    if (!running) {
      running = true
      try {
        if (configData.verbose) {
          console.log(`Autobarrel running due to ${p}`)
        }
        await autobarrel(configData)
      } catch (e) {
        console.error(e)
      } finally {
        running = false
      }
    }
  }
  console.log("Autobarrel watching...")
  watcher.on("add", (e) => {
    if (e.match(/\.tsx?$/)) {
      maybeAutobarrel(e)
    }
  })
  watcher.on("addDir", maybeAutobarrel)
  watcher.on("unlink", (e) => {
    if (e.match(/\.tsx?$/)) {
      maybeAutobarrel(e)
    }
  })
  watcher.on("unlinkDir", maybeAutobarrel)
  return watcher
}

export async function resolveAutobarrelConfig(config: AutoBarrelConfig) {
  if (!path.isAbsolute(config.path)) {
    throw new Error(
      `resolveAutobarrelConfig expects config.path to be absolute`
    )
  }
  const stat = await statAsync(config.path)
  if (!stat.isFile()) {
    throw new Error(`Autobarrel config ${config.path} is not a file`)
  }
  let configData: AutoBarrelData
  try {
    configData = JSON.parse(await readFileAsync(config.path, "utf8"))
  } catch (e) {
    if (e instanceof SyntaxError) {
      throw new Error(
        `Invalid syntax in the autobarrel JSON file: ${config.path}`
      )
    }
    throw e
  }

  if (!Array.isArray(configData.paths)) {
    throw new Error(`Missing required paths Array in autobarrel config`)
  }
  if (configData.exclude && !Array.isArray(configData.exclude)) {
    throw new Error(
      `Expected autobarrel exclude to be an array, saw ${configData.exclude}`
    )
  }
  if (configData.ignore && !Array.isArray(configData.ignore)) {
    throw new Error(
      `Expected autobarrel ignore to be an array, saw ${configData.exclude}`
    )
  }
  const absPathDir = path.dirname(config.path)

  Object.keys(configData).forEach((k) => {
    const key = k as keyof typeof configData
    if (Array.isArray(configData[key])) {
      configData[key] = configData[key]!.map((p) =>
        path.isAbsolute(p) ? p : path.join(absPathDir, p)
      )
    }
  })
  return configData
}
