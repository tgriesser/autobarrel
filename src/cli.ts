#!/usr/bin/env node
import fs from "fs"
import minimist from "minimist"
import path from "path"
import {
  autobarrel,
  autobarrelWatch,
  resolveAutobarrelConfig,
} from "./autobarrel"

//
const argv = minimist(process.argv.slice(2))

if (argv.h || argv.help) {
  console.log(`Syntax: autobarrel [options]\n`)
  console.log(`Example: autobarrel --config path/to/autobarrel.json --watch\n`)
  console.log(`Options:\n`)
  console.log(`--help       Print this message`)
  console.log(`--watch      Watch mode, via chokidar`)
  console.log(
    `--config     Specify the path to autobarrel.json config, default to cwd`
  )
  process.exit()
}

const confPath = argv.config
  ? path.isAbsolute(argv.config)
    ? argv.config
    : path.join(process.cwd(), argv.config)
  : path.join(process.cwd(), "autobarrel.json")

if (
  !fs.existsSync(confPath) &&
  !fs.existsSync(confPath.replace(path.basename(confPath), "tsconfig.json"))
) {
  throw new Error(
    `Could not find autobarrel.json file at ${confPath}, specify a valid path via --config`
  )
}

;(async function runAutoBarrel() {
  const configData = await resolveAutobarrelConfig({ path: confPath })
  if (argv.watch) {
    await autobarrelWatch(configData)
  } else {
    await autobarrel(configData)
    console.log(`Autobarrel complete for ${configData.paths}`)
  }
})().catch((e) => {
  console.error(e)
  process.exit(1)
})
