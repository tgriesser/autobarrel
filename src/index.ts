// created by autobarrel, do not modify directly

export * from './autobarrel'