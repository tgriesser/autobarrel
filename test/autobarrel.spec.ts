import * as glob from "glob"
import { autobarrel, resolveAutobarrelConfig } from "../src/autobarrel"
import fs from "fs-extra"
import path from "path"

describe("autobarrel", () => {
  beforeEach(() => {
    try {
      fs.rmdirSync(path.join(__dirname, "testing"), { recursive: true })
    } catch {}
    fs.copySync(
      path.join(__dirname, "fixtures"),
      path.join(__dirname, "testing")
    )
  })

  afterEach(() => {
    try {
      fs.rmdirSync(path.join(__dirname, "testing"), { recursive: true })
    } catch {}
  })

  test("autobarrel tests", async () => {
    const pathsWritten = await autobarrel(
      await resolveAutobarrelConfig({
        path: path.join(__dirname, "testing", "autobarrel.json"),
      })
    )
    const globFiles = glob.sync("testing/**", {
      cwd: __dirname,
    })
    expect(globFiles).toMatchSnapshot()
    const indexFiles = globFiles.filter((f) => f.endsWith("index.ts"))

    indexFiles.map((f) => {
      expect(fs.readFileSync(path.join(__dirname, f), "utf8")).toMatchSnapshot(
        f
      )
    })

    expect(pathsWritten).toMatchSnapshot("Paths Written")
  })
})
