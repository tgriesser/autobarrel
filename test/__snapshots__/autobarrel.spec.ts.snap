// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autobarrel autobarrel tests 1`] = `
Array [
  "/Users/tgriesser/Github/oss/autobarrel/test/testing",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/autobarrel.json",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/README.md",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedDir",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedDir/excluded.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedDir/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedFile",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedFile/excludedFile.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedFile/includedFile.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedFile/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedPattern",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedPattern/abc.bin.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedPattern/def.bin.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedPattern/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedPattern/some.gen.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludeDts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludeDts/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludeDts/nonDts.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludeDts/something.d.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/ignoredDir",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/ignoredDir/ignoredDir.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/jsTs",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/jsTs/file.js",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/jsTs/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/jsTs/tsFile.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/noTsFiles",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/noTsFiles/README.md",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/path",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/path/file.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/path/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/someDir",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/someDir/app.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/someDir/index.ts",
]
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludeDts/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './nonDts'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedDir/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './excluded'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedFile/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './includedFile'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedPattern/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './some.gen'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './excludeDts/'
export * from './excludedFile/'
export * from './excludedPattern/'
export * from './jsTs/'
export * from './some/'
export * from './someDir/'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/jsTs/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './tsFile'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './path/'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/path/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './file'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './deep/'"
`;

exports[`autobarrel autobarrel tests: /Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/someDir/index.ts 1`] = `
"// created by autobarrel, do not modify directly

export * from './app'"
`;

exports[`autobarrel autobarrel tests: Paths Written 1`] = `
Array [
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludeDts/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedDir/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedFile/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/excludedPattern/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/jsTs/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/deep/path/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/some/index.ts",
  "/Users/tgriesser/Github/oss/autobarrel/test/testing/testDir/someDir/index.ts",
]
`;
